fedimint-server/src/multiplexed.rs

```diff
@@ -134,28 +135,33 @@ where
                 // since we've last checked.
                 drop(out_of_order);
 
-                let (peer, new_msg) = connections.receive().await?;
-                if key == new_msg.key {
-                    return Ok((peer, new_msg.msg));
-                }
+                // Avoid holding the connections lock which prevents sends
+                // FIXME: should probably spawn a task for receiving instead
+                if let Some(Ok((peer, new_msg))) = connections.receive().now_or_never() {
```

Comment:
This is kind of a hack.

Architecturally the problem is that we're putting "sending" and "receiving" on the same object, which gets locked by the same lock. Before parallel DKG change I guess nowhere in the code anything expected that sending and receiving can be happening at the same time.

This is similar to e.g. sending and receiving on IO, which then get split by https://docs.rs/tokio/latest/tokio/io/fn.split.html , because they are in fact independent and happen in parallel, so should not be locked by the same thing.


The:

```
pub trait IPeerConnections<Msg>
where
    Msg: Serialize + DeserializeOwned + Unpin + Send,
{
```

should become two traits and the consensus code should hold two separate handles -one for sending, one for receiving. Under the hood we can keep a single IO thread handling each peer.

Alternatively, we could just make `send` and `recv` on `IPeerConnections` take `&self` instead of `&mut` so it wouldn't be necessary to take a lock outside of it, and turn:

```
pub struct ReconnectPeerConnections<T> {
    connections: HashMap<PeerId, PeerConnection<T>>,
}


struct PeerConnection<T> {
    outgoing: Sender<T>,
    incoming: Receiver<T>,
}
```

into:

pub struct ReconnectPeerConnections<T> {
    senders: Mutex<HashMap<PeerId, Sender<T>>>,   
    receivers: Mutex<HashMap<PeerId, Receiver<T>>>,
}
```

so we can lock sending and receiving part independently.

@elsirion what do you think?

Reply:
I don't think we need mutexes even. There could be a fn `PeerConnections::split(&'a mut self) -> (ReadHalf<'a>, WriteHalf<'a>)` that separately borrows the parts of the connections struct. The impl for `ReconnectPeerConnections` could return a `HashMap<PeerId, &Sender<T>>` and `HashMap<PeerId, &mut Receiver<T>` for read/write respectively and impl send/receive on these.

Reply:
@elsirion Something tells me this is going to difficult to implement and work with due to everything being borrowed, lifetimes, etc.

Right now, since methods on `IPeerConnections` take `&mut self`, they force wrapping in `Arc<Mutex<...>>` if anyone wants to use `IPeerConnections` concurrently (which we now do if we want parallelized DKG). So we effectively do have a `Mutex` already, just on the user's side.

Putting this `Mutex` inside just takes the responsibility of locking from the user to the implementation. And splitting it into two, just enables parallel write and read.

Reply:
The reason IO goes with `split` (despite much worse ergonomics) must be that operations on each end (read / write) don't make much sense with just `&self`, as they are inherently sequential, so no one should want `fn read(&self, &mut [u8])` that they can call in parallel. Not only it makes implementing such interfaces harder/slower, but it's just not useful for the caller. What the caller wants is `fn read(&mut self)` and `fn write(&mut self)`... just on a separate object.

But on peer connection, it does make sense as each message is in itself atomic. Multiple parallel `send`s just doesn't guarantee ordering (which the receiver might just not care about anyway), while multiple parallel `recv`... well.. technically works (again - only loosing ordering), though doesn't seem all that useful functionally in practice, I guess.

I guess `IPeerConnections::split` could be OK, under the condition that that it is not borrowing, but taking `self` by move.  

But then... why even bother with splitting. Just start with two separate objects to begin with, like I mentioned here:

> should become two traits and the consensus code should hold two separate handles -one for sending, one for receiving.

Reply:
I looked into splitting, there's not a very clean way to do it without splitting `ReconnectPeerConnections`.

Instead I removed all the `Mutex` and `sleep` and replaced it with channels, which greatly improved performance (see updated metrics in my first comment).

Reply:
> I looked into splitting, there's not a very clean way to do it without splitting ReconnectPeerConnections.

We should split it, yes.

> Instead I removed all the Mutex and sleep and replaced it with channels, which greatly improved performance in some cases.

Looks like quite a bit overhead, turning it into callback-based rpc system. Not excited about it, though I guess it works. Intuition tells me that 1000-sized bounded channels might lead to deadlocks, through probably only theoretically.

That 10ms sleep there leads to unnecessary delays, especially now that the contention in that place happens more oftent. Thats a good point/find.

Reply:
> That 10ms sleep there leads to unnecessary delays, especially now that the contention in that place happens more oftent. Thats a good point/find.

I've put a comment there, that it ideally should be a CondVar-based wait.

Reply:
> if anyone wants to use `IPeerConnections` concurrently

But nobody is supposed to use the read part of peer connections concurrently, this will always lead to weird bugs imo. I think we want a solution for allowing parallel writes though. But it sounds like kitman found another solution, taking a look at that.